File: downloads/code/LeetCode/Python/Binary-Tree-Preorder-Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        ret = []
        self.searchTree(root,ret)
        return ret
        
    def searchTree(self, root, ret):
        if root == None:
            return
        ret.append(root.val)
        if root.left != None:
            self.searchTree(root.left,ret)
        if root.right!=None:
            self.searchTree(root.right,ret)
        return

File: downloads/code/LeetCode/Java/Binary-Tree-Preorder-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ret = new LinkedList<Integer>();
        if(root==null)
            return ret;
        Stack<TreeNode> rights = new Stack<TreeNode>();
        
        while(root!=null){
            ret.add(root.val);
            if(root.right!=null)
                rights.push(root.right);
            
            root = root.left;
            if (root==null && !rights.isEmpty())
                root = rights.pop();
        } 
        return ret;   
    }
}
